10 決済情報機能

terukyou opened the pull request:
基本設計書のp34に下記二つの変更を加えました。
* クレジットカードは2件まで登録できます」のエラー表示は、
削除ボタンではなく、新規会員登録ボタンを押したときに表示するようにする。
* 新規会員登録ボタンは2件登録していると押せないようにする。

hirayama-sachi committed:
add: 決済情報機能初回作成

kazuyanakata commented on html/mycakeapp/src/Model/Table/CreditcardsTable.php:
```diff
@@ -55,31 +55,80 @@ public function initialize(array $config)
+                'rule' => ['HalfSizeAlphabet'],
+                'provider' => 'custom',
+                'message' => '半角英字以外の文字が使われています'
+            ]);
```
passwordに関連するところはないので、ここは不要だと思います！

terukyou replied:
確認ミスです、失礼しました💦
修正します！

kazuyanakata commented on html/mycakeapp/src/Model/Table/CreditcardsTable.php:
```diff
@@ -55,31 +55,80 @@ public function initialize(array $config)
             ->integer('deadline')
             ->requirePresence('deadline', 'create')
-            ->notEmptyString('deadline');
+            ->notEmptyString('deadline')
```
notEmptyStringが重複しています！最初の方は不要なので削除お願いします！

terukyou replied:
こちらも確認ミスです、申し訳ない。
修正します。

kazuyanakata commented on html/mycakeapp/src/Model/Table/CreditcardsTable.php:
```diff
@@ -55,31 +55,80 @@ public function initialize(array $config)
+        $validator
+            ->integer('security_code')
+            ->requirePresence('security_code', true, '空白になっています')
+            ->notEmptyString('security_code', '空白になっています')
```
notEmpryStringが重複してます！削除お願いします！

kazuyanakata commented on html/mycakeapp/src/Model/Table/CreditcardsTable.php:
```diff
@@ -55,31 +55,80 @@ public function initialize(array $config)
             ->maxLength('name', 100)
             ->requirePresence('name', 'create')
-            ->notEmptyString('name');
+            ->notEmptyString('name')
```
notEmpryStringが重複してます！削除お願いします！

kazuyanakata commented on html/mycakeapp/src/Model/Table/CreditcardsTable.php:
```diff
@@ -55,31 +55,80 @@ public function initialize(array $config)
+                'rule' => ['HalfSizeAlphabetOnly'],
+                'provider' => 'custom',
+                'message' => '半角英字以外の文字が使われています'
+            ])
```
姓と名の間に空白を入れたら保存ができません。
修正お願いします。

terukyou replied:
その通りですね..!!
修正します!!

kazuyanakata commented on html/mycakeapp/src/Controller/MypageController.php:
```diff
@@ -42,4 +45,138 @@ public function delete()
+            }
+        }
+        if (!empty($this->request->is('post'))) { //insert
+            $entity = $this->Creditcards->patchEntity($entity, $this->request->getData());
```
クレジットカード番号が暗号化されずに保存されています。
DB定義書ではクレジットカードは暗号化された状態で保存することになっているので対応お願いします。
(ここでクレジットカードがencryptされるので表示の際にはdecryptが必要だと思うので、そこも合わせて修正お願いします！)

terukyou replied:
暗号化はできたのですが、実現するにはクレジットカード番号の型をvarbinaryにする必要があるので、また共有させていただきます。

kazuyanakata commented on html/mycakeapp/src/Controller/MypageController.php:
```diff
@@ -42,4 +45,138 @@ public function delete()
+        $securitySalt = Configure::read('salt');
+        foreach ($cardsInfoOwn as $key => $value) {
+            $lastFourDigits = substr($value['card_number'], -4, 4);
+            $encryptedCardId = Security::encrypt($value['id'], $securityKey, $securitySalt);
```
<GOOD POINT!!!!!!>(修正じゃないです笑)
idを暗号化するのめっちゃいいと思いました！！さすがです！！

terukyou replied:
ありがとうございます！！🎉🤸‍♀️

kazuyanakata commented on html/mycakeapp/webroot/js/checkpayment.js:
```diff
@@ -0,0 +1,51 @@
+      });
+    });
+  };
+  $('.add-payment').on('click', (e) => {
+    if ($('.add-payment').hasClass('pressed')) { //無効ボタン化済みの時は↓の.add-paymentイベントは反応させない
+      e.preventDefault
+      e.stopImmediatePropagation();
+    };
+  });
+  if (paymentButtons.length !== 2) {
+    $('.add-payment').on('click', () => {
+      location.href = 'http://localhost:10380/mypage/addpayment';
+    });
+  } else {
+    $('.add-payment').on('click', () => {
+      $('.add-payment').addClass('pressed'); //ボタンを無効化するクラス
+      const addPaymentError = getElement('<p>', 'add-payment-error', 'クレジットカードは2件まで登録できます');
+      $('.content').append(addPaymentError);
+    });
+  }
+});
```
※修正不要
一度ボタンを押してエラー画面が出てくるので基本設計上全く問題ないのですが、
そもそも新規会員登録ボタンを2個登録してる時は表示しないとかボタンの色を押せない感じに変更するっていう設計でもいいなと思いました！
あくまで個人の感想なので修正不要です！

kazuyanakata commented on html/mycakeapp/webroot/js/checkpayment.js:
```diff
@@ -0,0 +1,51 @@
+  } else {
+    $('.add-payment').on('click', () => {
+      $('.add-payment').addClass('pressed'); //ボタンを無効化するクラス
+      const addPaymentError = getElement('<p>', 'add-payment-error', 'クレジットカードは2件まで登録できます');
```
（提案）
ここのエラー分、このままだと2件まで登録できるから登録してねって意味に捉えられるかもしれないので、
「登録できるカードの枚数は2枚までです」
ってするのはどうでしょうか？

terukyou replied:
確かにそちらの方が意図が伝わりやすいですね....
分かりました！修正します！！

kazuyanakata commented on html/mycakeapp/src/Controller/MypageController.php:
```diff
@@ -42,4 +45,138 @@ public function delete()
+        if (empty($_SESSION['addedpayment'])) {
+            return $this->redirect(['controller' => 'error']);
+        }
+        $_SESSION['addedpayment'] = null;
```
この形だと$_SESSIONに['addedpayment']が値無しの状態残ってしまうので、
```suggestion
        $this->request->session()->delete('addedpayment');
```
の方が良いと思います！

terukyou replied:
ありがとうございます！参考になります！！
こちらも修正いたします。

kazuyanakata commented on html/mycakeapp/src/Controller/MypageController.php:
```diff
@@ -42,4 +45,138 @@ public function delete()
+        if (empty($_SESSION['deletedpayment'])) {
+            return $this->redirect(['controller' => 'error']);
+        }
+        $_SESSION['deletedpayment'] = null;
```
上に同じくです

kazuyanakata commented on html/mycakeapp/src/Controller/MypageController.php:
```diff
@@ -42,4 +45,138 @@ public function delete()
+            }
+
+            $entity = $this->Creditcards->get($cardId);
+            $entity = $this->Creditcards->patchEntity($entity, $this->request->getData());
+            //updateの日付のみ更新
+            if ($this->Creditcards->save($entity)) {
+                $_SESSION['addedpayment'] = 1;
+                return $this->redirect(['action' => 'addedpayment']);
+            }
```
編集画面にいくと、セキュリティコードとチェックボックスにエラー文が最初からついてしまっているので対応お願いします。(恐らく原因はここなのですが、私もここの対応方法分かっていないので、同時進行で考えるって感じにしましょう！笑。もし私の方で解決方法が出たら、また共有します！)

terukyou replied:
対処法一度考えていただいてありがとうございます...!!
確かにそうですね..。一度、自分で考えて修正します。

kazuyanakata commented on html/mycakeapp/src/Controller/MypageController.php:
```diff
@@ -42,4 +45,138 @@ public function delete()
+                return $this->redirect(['action' => 'addedpayment']);
+            }
+        } else { //insertのカード数判定
+            $numbeOfCardsOwned = $this->Creditcards->find()->where([
```
```suggestion
            $numberOfCardsOwned = $this->Creditcards->find()->where([
```
恐らくタイポかと思われます。

terukyou replied:
すみません、タイポです..
修正します！！

kazuyanakata commented on html/mycakeapp/src/Controller/MypageController.php:
```diff
@@ -42,4 +45,138 @@ public function delete()
+                'is_deleted' => 0,
+                'deadline >=' => $today
+            ])->count();
+            if ($numbeOfCardsOwned === 2) {
```
```suggestion
            if ($numberOfCardsOwned === 2) {
```
上に同じくタイポかと

kazuyanakata reviewed (commented):
レビュー完了しました！
処理の流れがシンプルで読みやすかったし、idを暗号化して渡してるの良いと思いました！！
数件コメント入れたので確認お願いします！！

hirayama-sachi committed:
Merge branch 'develop' of https://github.com/labotinc/codegym-team5 into feature/hinappu-10/#61

hirayama-sachi committed:
fix: 不要なバリデーションを削除

CreditcardsTableのnotEmpryStringと同テーブルのpasswordの削除

hirayama-sachi committed:
fix: typoの修正

$numberOfCardsOwnedへ修正

hirayama-sachi committed:
fix: URL直打ち対策のためのセッションの削除方法を変更

nullではセッションが残ってしまうため、Session::delete使用するよう変更

hirayama-sachi committed:
improve: クレジットの登録数を制限するエラーメッセージの文言を変更

メッセージの意図が分かりやすくなるよう変更

terukyou commented:
> passwordに関連するところはないので、ここは不要だと思います！ 

失礼しました💦
修正します！

hirayama-sachi committed:
fix: 最初のカードの編集画面にエラー表示が出ないよう修正

putで送信された際のみvalidationが通るよう修正

hirayama-sachi committed:
fix: update時にupdate_atが更新されていなかったので修正

hirayama-sachi committed:
fix: クレジットカード番号を暗号化してDBに保存し表示する際は復号化するよう修正

独自タイプを作成し、暗号化復号化を実現

hirayama-sachi committed:
fix: カードの名義人入力フォームを半角の空白を含めた半角英字を許可するバリデーションへ変更

半角英字と半角のスペースのみ許可するよう変更した

terukyou commented:
**ご指摘いただいたもの**
74015ca passwordのバリデーション削除
74015ca notEmptyStringの重複を削除
3277e8e 姓と名の間の空白を許可するバリデーションへ変更
00f7ebc クレジットカード番号の暗号化
0953084 カードの登録数を制限するエラーの文を変更
cf371b8 セッションをSession::deleteで削除するよう変更
c9b81dd カードの編集画面(get)の時バリデーションエラーが表示されないよう修正
c126cb8 タイポの修正
**修正の際見つけたバグ**
583221d update時にupdate_atが更新されていなかったので修正

hirayama-sachi committed:
Merge branch 'develop' of https://github.com/labotinc/codegym-team5 into feature/hinappu-10/#61

kazuyanakata commented on html/mycakeapp/src/Model/Table/CreditcardsTable.php:
```diff
@@ -102,8 +142,26 @@ public function validationDefault(Validator $validator)
+                    return true;
+                }
+                //  card_number と deleted の組み合わせで重複チェック
+                $rule = $rules->isUnique(
+                    ['card_number', 'is_deleted'],
+                    'このクレジットカードは利用できません'
+                );
+                return $rule($entity, $options);
```
おそらく、クレジットカードの暗号化が原因で同じカード番号でも重複して登録できるようになっております。
対応お願いいたします。

kazuyanakata commented on html/mycakeapp/src/Controller/MypageController.php:
```diff
@@ -42,4 +45,140 @@ public function delete()
+                return $this->redirect(['action' => 'addedpayment']);
+            }
+        }
+        $title = "会員登録";
```
タイトルが会員登録になっているので「決済情報」に変更お願いします。

terukyou replied:
失礼しました💦
修正します...

kazuyanakata commented on html/mycakeapp/src/Controller/MypageController.php:
```diff
@@ -42,4 +45,140 @@ public function delete()
+            return $this->redirect(['controller' => 'error']);
+        }
+        $this->request->session()->delete('addedpayment');
+        $title = "会員登録";
```
上に同じくです

kazuyanakata commented on html/mycakeapp/src/Controller/MypageController.php:
```diff
@@ -42,4 +45,140 @@ public function delete()
+            return $this->redirect(['controller' => 'error']);
+        }
+        $this->request->session()->delete('deletedpayment');
+        $title = "会員登録";
```
上に同じくです

hirayama-sachi committed:
fix: 決済登録と登録完了ページのフレームタイトルの誤字修正

kazuyanakata commented on html/mycakeapp/src/Model/Table/CreditcardsTable.php:
```diff
@@ -55,31 +56,70 @@ public function initialize(array $config)
             ->integer('deadline')
             ->requirePresence('deadline', 'create')
-            ->notEmptyString('deadline');
+            ->integer('deadline', '半角数字以外の文字が使われています')
```
先ほどディスコードでもお話しした通り、有効期限は「MM/YY」という形で入力させるようにしましょう。なので、それに関するバリデーションの設定をお願いいたします。
また、現在日付としてありえない日も日付として設定できてしまうので、そこの制限も合わせて修正お願いいたします。ex)14月35日とかも入力可能になっている。

kazuyanakata reviewed (commented):
レビュー完了しました！
確認お願いいたします！

hirayama-sachi committed:
fix: controllerから重複チェックを行うよう修正

暗号化をしたためmodelの重複ルールチェックが効かなくなり、その対処としてcontrollerから重複判定するようにした

hirayama-sachi committed:
Merge branch 'develop' of https://github.com/labotinc/codegym-team5 into feature/hinappu-10/#61

hirayama-sachi committed:
fix: フレームタイトルの誤字修正

削除完了画面のフレームタイトルの誤字があったので修正した

hirayama-sachi committed:
fix: 有効期限をmm/yyに変更、それに伴いfind、validation、viewの変更

クレジットカードの記載通りに入力可能、実在する未来の日付か判定可能、findが
冗長になったためテーブルに以降しコントローラ―がリファクタリングされた
findの参考)https://qiita.com/azukiazusa/items/3c4b2d6181cf026bd02e

hirayama-sachi committed:
fix: カード番号が一意でないのに2回送信すると通ってしまうバグをif文の順を変えることで対処

hirayama-sachi committed:
fix: カード名義のカスタムバリデーション(空白)が通っていなかったので修正

ruleNameに各々のカスタム名を付与することで対処した

hirayama-sachi committed:
Merge branch 'develop' of https://github.com/labotinc/codegym-team5 into feature/hinappu-10/#61

terukyou commented:
**ご指摘いただいたもの**
18dddac クレジットカードの重複チェック
0b97bef 84beceb タイトルの誤字
680ee1d 有効期限をMM/YYに変更
**修正の際見つけたバグ**
4f82aaf カード番号が一意でないのに2回送信すると通ってしまうバグの修正
62507e3 カード名義のカスタムバリデーション(空白)が通らないバグの修正

kazuyanakata commented on html/mycakeapp/src/Controller/MypageController.php:
```diff
@@ -42,4 +45,135 @@ public function delete()
+            $entity = $this->Creditcards->patchEntity($entity, $this->request->getData());
+
+            $isCardNumberExist = in_array($entity['card_number'], $this->Creditcards->find('AllCardNumbers'), true);
+            if ($isCardNumberExist === true && $entity['card_number'] !== $previousCardNumber) {
+                $cardNumberIsNotUnique = "このクレジットカードは利用できません";
+                $this->set(compact('cardNumberIsNotUnique'));
+            }
+            $entity["updated_at"] = date("Y/m/d H:i:s");
+            if (empty($cardNumberIsNotUnique) && $this->Creditcards->save($entity)) {
+                $_SESSION['addedpayment'] = 1;
+                return $this->redirect(['action' => 'addedpayment']);
+            }
+        }
```
今のままだと、同じクレジットカード番号を登録したときに、エラー文は表示されるがDBに登録されてしまう状態になっています。
```suggestion
            if ($isCardNumberExist === true && $entity['card_number'] !== $previousCardNumber) {
                $cardNumberIsNotUnique = "このクレジットカードは利用できません";
                $this->set(compact('cardNumberIsNotUnique'));
            } else {
                $entity["updated_at"] = date("Y/m/d H:i:s");
                if (empty($cardNumberIsNotUnique) && $this->Creditcards->save($entity)) {
                    $_SESSION['addedpayment'] = 1;
                    return $this->redirect(['action' => 'addedpayment']);
                }
            }
```
こうすることで、エラー文があるときにsaveを通ることがなくなるので対応お願いいたします！

kazuyanakata commented on html/mycakeapp/src/Controller/MypageController.php:
```diff
@@ -42,4 +45,135 @@ public function delete()
+        if (!empty($this->request->is('post'))) { //insert
+            $entity = $this->Creditcards->patchEntity($entity, $this->request->getData());
+            $isCardNumberExist = in_array($entity['card_number'], $this->Creditcards->find('AllCardNumbers'), true);
+            if ($isCardNumberExist === true) {
+                $cardNumberIsNotUnique = "このクレジットカードは利用できません";
+                $this->set(compact('cardNumberIsNotUnique'));
+            }
+            $entity["member_id"] = $this->Auth->user('id');
+            $entity["is_deleted"] = 0;
+            $entity["created_at"] = date("Y/m/d H:i:s");
+            $entity["updated_at"] = date("Y/m/d H:i:s");
+            if ($this->Creditcards->save($entity) && empty($cardNumberIsNotUnique)) {
+                $_SESSION['addedpayment'] = 1;
+                return $this->redirect(['action' => 'addedpayment']);
+            }
```
上に同じくです
```suggestion
            if ($isCardNumberExist === true) {
                $cardNumberIsNotUnique = "このクレジットカードは利用できません";
                $this->set(compact('cardNumberIsNotUnique'));
           } else {
                $entity["member_id"] = $this->Auth->user('id');
                $entity["is_deleted"] = 0;
                $entity["created_at"] = date("Y/m/d H:i:s");
                $entity["updated_at"] = date("Y/m/d H:i:s");
                if ($this->Creditcards->save($entity) && empty($cardNumberIsNotUnique)) {
                    $_SESSION['addedpayment'] = 1;
                    return $this->redirect(['action' => 'addedpayment']);
                }
            }
```

kazuyanakata commented on html/mycakeapp/src/Model/Table/CreditcardsTable.php:
```diff
@@ -107,4 +151,52 @@ public function buildRules(RulesChecker $rules)
+        foreach ((array)$allCardNumbersArray as $allCardNumberArray) {
+            $deadlineInfo = date_parse_from_format('m/y', $allCardNumberArray->deadline);
+            $deadline = $deadlineInfo['year'] . $deadlineInfo['month'];
+            if ((int)$deadline > (int)$today) {
```
有効期限が今月のカードもまだ使えるので、今月分は含めた方がいいと思います！
```suggestion
            if ((int)$deadline >= (int)$today) {
```

kazuyanakata commented on html/mycakeapp/src/Model/Table/CreditcardsTable.php:
```diff
@@ -107,4 +151,52 @@ public function buildRules(RulesChecker $rules)
+        foreach ((array)$registeredCardsInfoOwn as $registeredCardInfoOwn) {
+            $deadlineInfo = date_parse_from_format('m/y', $registeredCardInfoOwn->deadline);
+            $deadline = $deadlineInfo['year'] . $deadlineInfo['month'];
+            if ((int)$deadline > (int)$today) {
```
上と同じ理由です
```suggestion
            if ((int)$deadline >= (int)$today) {
```

kazuyanakata commented on html/mycakeapp/src/Model/Validation/CustomValidation.php:
```diff
@@ -17,4 +17,27 @@ public function HalfSizeAlphanumericOnly($value)
+        if ($month > 12 || $year < $thisYear) {
+            return false;
+        }
+        if ($month <= $thisMonth && $year === $thisYear) {
```
上と同じ理由です。今月分の登録は許可した方がいいと思います！
```suggestion
        if ($month < $thisMonth && $year === $thisYear) {
```

kazuyanakata commented on html/mycakeapp/src/Controller/MypageController.php:
```diff
@@ -42,4 +45,135 @@ public function delete()
+    {
+        $this->viewBuilder()->setLayout('frame-title');
+        $memberId = $this->Auth->user('id');
+        $cardsInfoOwn = $this->Creditcards->find('CardsInfoOwn', ['memberId' => $memberId]);
```
<Good Point!!!>
findCardInfoOwnでやるやり方、コントローラーのコーディングもシンプルになるし、汎用的に使えるのでとても良いと思いました！！
真似します！！笑

kazuyanakata reviewed (commented):
レビュー完了しましたので確認お願いいたします！

hirayama-sachi committed:
fix: カードの有効期限を今月も許可するよう変更

今月のカードはまだ使用可能なので、有効期限が今月のものは登録・編集・表示が可能になるよう修正した

hirayama-sachi committed:
fix: カード番号が重複する場合saveは通らないように修正

カード番号が重複した時、エラーは出るがDBに保存される可能性があっため、エラーメッセージが出た時はsaveが通らないようより厳密にした

kazuyanakata reviewed (commented):
LGTMです！！！

全体を通して、処理が長くなっても(これは機能上しょうがないことだと思うので気にせず)変数名や単体で見た時の処理が分かりやすかったので、レビューしててとても勉強になりました！！

マージお願いします！
